### ISSUE TITLE
Deleting a volume will not delete its PV/PVC. 

### ISSUE THREAD

**Issue by meldafrawi** [longhorn/longhorn#520]
- Using UI:
  - Create a volume.
  - Create PV/PVC
  - Delete the volume.
- Using `kubectl get pv`
  - PV/PVC not deleted.

**Comment by yasker** [comment]
We need to think more about deleting a volume.

Currently deleting a volume won't be blocked by any operations.
1. PV can be deleted while deleting the volume. But PVC is the declaration of a workload. User can manually create it for its workload. I don't think we can just delete it.
2. If there is workload current using the volume, should we allow deleting the volume? Or, should we even allow detach the volume?

**Comment by yasker** [comment]
keep it open

**Comment by ttpcodes** [comment]
@joshimoo and I had a discussion regarding handling of `Volume` deletion in various cases in which a `PersistentVolume` and `PersistentVolumeClaim` are involved. This is what we eventually decided on:
- It's possible that there are other inactive workloads such as a `Deployment` scaled to zero `Pods` that depend on the `Volume`. While we'd be able to successfully delete the `PersistentVolume` and `PersistentVolumeClaim`, it would cause problems once the `Deployment` scales back up and discovers that the `PersistentVolumeClaim` and other resources that it depends on no longer exists. We determined that it might be best to warn the user before deletion that there is a `PersistentVolume` and possibly `PersistentVolumeClaim` related to the `Volume` and  to be aware in case that it affects their workloads. Afterwards, they can delete the `Volume` with the option to have `Longhorn` handle the `PersistentVolume` and `PersistentVolumeClaim` deletion for them.
- Deleting a `Volume` with running workloads would cause obvious problems for the end user, and we should at the very least, block a user from deleting a `Volume` that has a workload attached.

I think it's safe to go ahead with working on the second case and preventing deletion of `Volumes` with associated workloads. As for the first case, @yasker do you have any additional thoughts on how this should be handled? Or does this look good and can we proceed with implementing it?

**Comment by yasker** [comment]
@ttpcodes 

I've thought about this a bit more.

1. We should know if the volume is referred by a PV and PVC, since any reference to the PVC will be written in the `claimRef` in the PV. We also know if the workload using the volume exists or not. We can provide information to the user, and make sure user checks the confirmation checkbox before deleting the volume.
2. Preventing the user from deleting an arbitrary volume is kind of preventing him from shooting himself in the foot. But if he insists on doing so, we shouldn't stand in the way. Just make sure we give a big enough warning for it.

Another thing I have a bit concern is, deleting a PV/PVC created from Longhorn UI should be straightforward, but how about PV/PVC created using StorageClass? There might be some complications there. We need that in the test plan.

Also, how are we going to implement this on the API level? We need some designs.

Can you prepare a Longhorn Enhancement Proposal for this? Thanks.

**Comment by ttpcodes** [comment]
## Pre-merged Checklist
* [X] Does the PR include the explanation for the fix or the feature?

* [X] Is the backend code merged (Manager, Engine, Instance Manager, BackupStore etc)?
The PR is at longhorn/longhorn-manager#1528.

* [X] Is the reproduce steps/test steps documented?

* [ ] Which areas/issues this PR might have potential impacts on?
Area: kubernetes, manager
Issues: #520

* [X] ~~**If the fix introduces the code for backward compatibility** Has an separate issue been filed with the label `release/obsolete-compatibility`?~~

* [ ] **If labeled: area/ui** Has the UI issue filed or ready to be merged?
The UI issue/PR is at longhorn/longhorn-ui#301.

* [ ] **if labeled: require/doc** Has the necessary document PR submitted or merged?
The Doc issue/PR is at 

* [ ] **If labeled: require/automation-e2e** Has the end-to-end test plan been merged? Have QAs agreed on the automation test case?
The automation skeleton PR is at
The automation test case PR is at

* [X] ~~**if labeled: require/automation-engine** Has the engine integration test been merged?~~

* [ ] **if labeled: require/manual-test-plan** Has the manual test plan been documented?
The updated manual test plan is at

**Comment by ttpcodes** [comment]
A number of cases need to be manually tested in the `longhorn-ui`. To test these cases, create the `Volume` with the specified conditions in each case, and then try to delete it. What is observed should match what is described in the test case:
1. A regular `Volume`. Only the default deletion prompt should show up asking to confirm deletion.
2. A `Volume` with a `Persistent Volume`. The deletion prompt should tell the user that there is a `Persistent Volume` that will be deleted along with the `Volume`.
3. A `Volume` with a `Persistent Volume` and `Persistent Volume Claim`. The deletion prompt should tell the user that there is a `Persistent Volume` and `Persistent Volume Claim` that will be deleted along with the `Volume`.
4. A `Volume` that is `Attached`. The deletion prompt should indicate what `Node` the `Volume` is attached to and warn the user about errors that may occur as a result of deleting an attached `Volume`.
5. A `Volume` that is `Attached` and has a `Persistent Volume`. The deletion prompt should contain the information from both test cases 2 and 4.

Additionally, here are bulk deletion test cases that need testing:
1. 1+ regular `Volumes`. Only the default deletion prompt should show up asking to confirm deletion.
2. 1+ `Volumes` with a `Persistent Volume`.  The deletion prompt should list the `Persistent Volumes` associated with the `Volumes` and tell the user that these will also be deleted.
3. 0+ regular `Volumes` and 1+ `Volumes` that are `Attached`. The deletion prompt should list only `Volumes` that are `Attached` and tell the user that applications using them may encounter errors once the `Volumes` are deleted.
4. 0+ regular `Volumes`, 1+ `Volumes` with a `Persistent Volume`, and 1+ `Volumes` that are `Attached`. The information described in test cases 2 and 3 should be displayed.

Finally, there are some other test cases to check here:
1. Create a `Volume` and create a `Persistent Volume` and `Persistent Volume Claim` through the `longhorn-ui`. Delete the `Persistent Volume Claim` manually. Delete the `Volume`. The deletion prompt should not list the `Persistent Volume Claim` that was deleted in the list of resources to be deleted.
2. Create a `Disaster Recovery Volume`. Delete the `Disaster Recovery Volume`. The deletion prompt should not give a warning about errors that may occur from deleting an attached `Volume`.
3. Create a `Volume` from a `Backup`. While the `Volume` is still being restored, delete the `Volume`. The deletion prompt should not give a warning about errors that may occur from deleting an attached `Volume`.

Note that not all of these tests will pass yet (since some points like bulk deletion case #3 are not addressed yet).

**Comment by meldafrawi** [comment]
Validation: **PASSED**

Steps (1/2):
- Create a volume, create a PV.
- Delete the volume, UI will warn about PV deletion, PV will get deleted.

Steps (2/2):
- Create a volume, create a PV/PVC
- Delete the volume, UI will warn about PV/PVC deletion, PV & PVC will get deleted.
